KEYCLOAK-9360 Two factor authentication with W3C Web Authentication - WIP 1st impl phase

tnorimat opened the pull request:
This PR is about W3C Web Authentication. This is the first implementation phase of the following documents :
JIRA Ticket :  https://issues.jboss.org/browse/KEYCLOAK-9360
Design Document : https://github.com/keycloak/keycloak-community/blob/master/design/web-authn-authenticator.md

This PR is still WIP and is intended for being reviewed at the early stage of implementation.

Therefore, this includes incomplete implementation and unresolved issues. I'll describe them as the following comments afterwards. I would be happy if this PR is reviewed.

tnorimat committed:
KEYCLOAK-9360 Two factor authentication with W3C Web Authentication - WIP 1st impl phase

tnorimat commented:
Public Key Credneital Management

Issue 1 : credential itself and credential's metadata are managed separately which make it hard to keep both of their lifecycle consistent

The former is managed via Credential Provider while the latter is managed as the attribute of UserModel.

Moreover, managing the credential's metadata as the attribute of UserModel seems to be inappropriate because those are exposed on Users->(user)->Attributes tab. The admin might unintentionally edit or delete them.

In my opinion, the credential's metadata is also managed via Credential Provider. However, the current Credential Providers can not have such method to do that.

tnorimat commented:
Testing

Issue 2 : the arquillian integration test including real browser instance(Chrome) can only test the basic process of user registration and authentication

Web Authentication Testing API is used for the test. However, it can not cover testing of all implemented features.

This testing API returns the same response at all times when calling WebAuthn API (navigator.credential.create(), .get()), which make us harder to write tests covering all implemented features.

tnorimat commented:
# Usage
***

## Preparation

### Enable User Registration

#### By Admin
On Admin Console
*  open Realm `Settings -> Login` tab
*  set `User registration` switch `ON` and push `Save` button

### Enable Webauthn Authenticator Registration

#### By Admin
On Admin Console
* open `Authentication -> Required Actions` tab
* push `Register` button
* select `Webauthn Register` as `Required Action`
* mark `Enabled` and `Default Action` checkbox

### Enable 2FA by Webauthn Authenticator

#### By Admin
On Admin Console
* open `Authentication -> Flows` tab
* copy and create the new browser flow
* add `WebAuthn Authenticator` execution below `User Password Form` execution
* set `REQUIRED` to `WebAuthn Authenticator` execution's requirement
* open `Authentication -> Bindings` tab
* select the created new browser flow as `Browser Flow`
* push `Save` button

## Register Webauthn Authenticator

### User without their account

#### By User
* open the login form
* click `Register` link
* fill in items on the register form and push `Register` button
* the user's browser asks the user to register their webauthn authenticator

### User with their account

#### By Admin
On Admin Console, disable authentication by webauthn authenticator tentatively
* open `Authentication -> Bindings` tab
* select the built-in browser flow as `Browser Flow`
* push `Save` button

require the user to register their webauthn authenticator after login
* open `Users -> (user) -> Details` tab
* set `Webauthn Register` on `Required User Actions`
* push `Save` button

#### By User
* open the login form
* fill in items and push `Login` button
* the user's browser asks the user to register their webauthn authenticator

#### By Admin
On Admin Console, re-enable authentication by webauthn authenticator
* open `Authentication -> Bindings` tab
* select the created new browser flow as `Browser Flow`
* push `Save` button

## Authenticate by Webauthn Authenticator
#### By User
* open the login form
* fill in the items and push `Login` button
* the list of registered webauthn authenticators are shown. push `Authenticate` button
* the user's browser asks the user to authenticate by their webauthn authenticator

## View Registered Webauthn Authenticator

The following information is shown :
- Public Key Credential ID
- Public Key Credential Label
- Public Key Credential AAGUID

#### By Admin
On Admin Console
* open `Users -> (user) -> Credentials` tab
* view `Manage WebAuthn Authenticator` area

#### By User
On User Account Service
* view `Account` page

## Edit Registered Webauthn Authenticator

The following information is editable :
- Public Key Credential Label

#### By Admin
 **Not yet realized**

#### By User
On User Account Service
* view `Account` page
* edit the value in `Public Key Credential Label`
* push `Save` button

## Delete Registered Webauthn Authenticator

#### By Admin
On Admin Console
* open `Users -> (user) -> Credentials` tab
* on `Disable Credentials` area, add `webauthn` to `Disable Types`
* push `Disable Credential Types` button

#### By User
**Not yet realized**

## Re-Register Webauthn Authenticator

Newly registered webauthn authenticator overrides the old registered one.

#### By Admin
On Admin Console
* open `Users -> (user) -> Details` tab
* set `Webauthn Register` on `Required User Actions`
* push `Save` button

#### By User
* open the login form
* fill in the items and push `Login` button
* the user's browser asks the user to register their webauthn authenticator

## Configuration

### Webauthn Registration

#### By Admin
On Admin Console
* open `Authentication -> WebAuthn Policy` tab
* configure items and push `Save` button


### Webauthn Authentication

#### by Admin
On Admin Console
* open `Authentication -> Flows` tab
* select the created new browser flow
* click Actions link of `WebAuthn Authenticator` execution and click `Config`
* configure items and push `Save` button

tnorimat commented:
Caching Public Key Credentials

Issue 3 : The publick key credential's Credential Provider has not yet implemented `OnUserCache` interface while other existing credentials' (password, OTP) Credential Providers have implemented. 

Therefore, the public key credential has not been cached.

Considering performance impact, The publick key credential's Credential Provider should also implement `OnUserCache` interface.

tnorimat commented:
Browser Side Implementation

Issue 4 : Considering secuirty and UX, there seems to be problems on the browser side implementation

 ### WebAuthn Registration - webauthn-register.ftl

#### How to retrieve parameters for WebAuthn API (nagivator.credentials.create())
- all parameters are embedded on the freemarker templete
or
- access REST API providing these parameters asynhchronously

#### UI for WebAuthn Authenticator Label edit
- window.prompt dialog
or
- other way
 
#### verifying and sanitizing against WebAuthn Authenticator Label value set by the user

### WebAuthn Authentication - webauthn-authenticate.ftl

#### How to retrieve parameters for WebAuthn API (nagivator.credentials.get())
- all parameters are embedded on the freemarker templete
or
- access REST API providing these parameters asynhchronously

#### UI for selecting registered webauthn authenticator
- table
or
- other way

tnorimat commented:
Edit Registered Webauthn Authenticator

Issue 5 : The registered webauthn authenticator can not editable by the admin.

tnorimat commented:
Delete Registered Webauthn Authenticator

Issue 6 : The registered webauthn authenticator can not deleted by the user.

tnorimat commented:
Security Consideration

Issue 7 : list up possible senarios in the real world and clarify whether there are some security flaw on each senario.

tnorimat commented:
That's all 7 issues I've recognized. I'll continue tackling them.

mposolda commented:
Thanks for the PR. This is great work!

Some more comments:

- Dependencies: I need to doublecheck with the team if it is possible to include all the new 3rd party dependencies. In addition to the 2 webauthn4j jars, there is also apache kerby and fasterxml-cbor. I will need to doublecheck if it's fine from the productization and other perspectives. If it's possible to remove some of those dependencies from the pull request (especially apache kerby), it will be best, but I guess you already checked it and it's not an easy option?

- One possible hint related to above: Keycloak has internally the class org.keycloak.common.util.Base64Url to encode/decode to Base64Url. So it may be possible to use this class to avoid some usage of Base64 from webauthn4j-utils? However I guess that still won't help as webauthn4j-utils will be required anyway for the other stuff and hence the 3rd party dependency will still be needed here? If it doesn't help, feel free to ignore this statement and still use the Base64 from webauthn-utils :)

- In the old account console ( http://localhost:8081/auth/realms/foo/account/ ) when I am authenticated as the user with the webauthn authenticator attached to my account, I am not able to edit the "Public Key Credential Label" . When I edit it to some value and click "Save", the label is not saved. I need to use admin console to change the label. TBH I would not spend much time with make the old account console CRUD working as the old account console will be decommissioned soon. Maybe just change the label to be read-only is the easiest workaround, so that all 3 labels are just read-only? Ideal would be if those 3 labels are shown just if webauthn authenticator is attached to the user account. From the longer term perspective, we will need to focus on the new account console.

- It will be nice if WebAuthn authenticator and requiredAction works in similar way as OTP authenticator and required action. This means:
-- WebAuthn authenticator can have requirements like "Required, Optional, Disabled" .
-- If WebAuthn authenticator is marked as "Optional" in the flow, it will be ignored in case that user doesn't have the webauthn authenticator already attached to him. But if user has webauthn attached, it will be required for user to authenticate with it. Currently it seems that when WebAuthn authenticator is marked "Alternative", it is just not used at all.
-- If WebAuthn authenticator is marked as "Required" in the flow and user doesn't have the webauthn authenticator attached, then authenticator shouldn't throw an error, but instead the requiredAction of "WebAuthn register" can be added to the user. In other words, user will be automatically added to the requiredAction without a need of admin to explicitly add requiredAction.
Currently when I want to test webauthn, I need to create authentication flow with WebAuthn authenticator disabled, then manually add requiredAction by admin to the user and after user registers the authenticator, I need to switch the authentication flow for the whole realm to have WebAuthn authenticator "Required" . This is not very ideal.

- Is it necessary to duplicate "userPreferred" configuration on both WebAuthnPolicy and Authenticator?

- Testing: Regarding testing, I think we can go with the ChromeDriver with enabled webauthn testing. We will need to figure later how to deal with testing webauthn configuration specific scenario. I think I will write email to keycloak-dev about it. However in the current PR, the RegistarAndLoginTest didn't pass on my laptop, so some more work is needed. Also will be good to avoid the dependency on some 3rd party library to manage WebDrivers as we're using arquillian drone for that. That can be done later and test can be possibly annotated with @Ignore to not block this PR IMO.

I've added some inline comments, but most of them are quite minor things. 

Besides that, feel free to continue on the issues you mentioned in the comment above, but IMO, the PR doesn't need to be perfect at the first stage. Still, I think that it will be merged at the end of August/start of September when we return from PTO...

Thanks again!

tnorimat commented:
@mposolda 
Thank you for your review and comments. I would like to resolve points you mentioned based on this PR. I also enter PTO soon and resume this work on the start of September when I return from PTO.

tnorimat committed:
credental label editable, authenticator support option, ignore test

tnorimat committed:
PlainTextVaultProvider incorporated

tnorimat commented:
#### 1. Dependencies (apache-kerby-asn1)

I hope that apache-kerby-asn1 is inculded at this 1st implementation phase.

At this 1st implementation phase, apache-kerby-asn1 is not used by the following reasons :

- apache-kerby-asn1 is used for treating Android Key Attestation Statement TPM Attestation Statement
- this 1st implementation phase does not verify any type of Attestation Statement

However, apache-kerby-asn1 is needed in the future when keycloak comes to support verifying Attestation Statement.

#### 2. Dependencies (webauthn4j-utils)

I hope webauthn4j-utils is inculded at this 1st implementation phase.

webauthn4j-utils' Base64URL encoder/decoder can be replaced with keycloak's ones. However, this PR uses `WebAuthnException` defined in webauthn4j-utils. webauthn4j-core and this PR uses `WebAuthnException` to notify that WebAuthn related error occurs. Also, webauthn4j-core itself uses some classes of webauthn4j-utils.

#### 3. WebAuthn Authenticator Label Edit on User Account Service

I've fixed to edit the label on the account page of User Account Service.

#### 4. The way of WebAuthn Authenticator working (similar to OTP Authenticator)

I see. I will fix WebAuthn Authenticator to work as you mentioned.

#### 5. User Verification Requirement configuration

I hope UserVerificationRequirement can be set on both WebAuthn authenticator's registration and authentication independently at this time.

UserVerificationRequirement can be configured independently according to WebAuthn specification.

If keycloak have one UserVerificationRequirement configuration, the following 3 types of configuration are possible :

- the same setting of it is applied to both WebAuthn authenticator's registration and authentication (e.g. "required" on registration and "required" on authentication ).

- the setting of it is applied to only WebAuthn authenticator's registration

- the setting of it is applied to only WebAuthn authenticator's authentication

I think that having one UserVerificationRequirement configuration seems to be valid if not any combination of different settings of UserVerificationRequirement (e.g. "required" on registration and "preferred" on authentication) other than the 3 ways of configuration above is meaningful in the real world. I'm not sure that.
 
#### 6. Testing

I've add @Ignore onto `RegisterAndLoginTest`.

tnorimat committed:
webauthn authenticator works the same as otp authenticator does

mposolda reviewed (commented):
ouch, I see that I somehow forgot to "confirm" the inline comments I added 2 weeks before :( So confirming now.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/webauthn/LoginPage.java:
```diff
@@ -0,0 +1,190 @@
+
+import static org.keycloak.testsuite.util.UIUtils.clickLink;
+
+public class LoginPage extends LanguageComboboxAwarePage {
```
Yes, I'll rename this LoginPage.java to WebAuthnLoginPage.java .

mposolda replied:
Thanks!

tnorimat commented on testsuite/integration-arquillian/tests/base/pom.xml:
```diff
@@ -44,6 +44,11 @@
     </properties>
     
     <dependencies>
+        <dependency>
```
I agree with you. I also would like to remove this dependency. 

I think it might be sufficient to realize the following:

* On conducting tests under webauthn package, use ChromeDriver and set "--enable-web-authentication-testing-api" option.

It is better to do that using exsinging arquillian/drone system. However, I'm not sure how to do it.

Therefore, I've implement the test codes by other way that actually run for the moment. 

As you advised, I'll look into AbstractX509AuthenticationTest.

mposolda replied:
This is definitely not a blocker for this PR IMO as I mentioned above.

We can create follow-up JIRA and someone from Keycloak team, who is more familiar with arquillian/drone can take a look into it. For now, the test is @Ignored anyway in the 1st stage of the PR, so I prefer to have something working in the keycloak master ASAP and then figure those details later :)

tnorimat replied:
I see. I'll work other tasks first.

tnorimat replied:
I've removed this dependency of Web Driver manager and use Arquillian Drone managing Chrome Driver. If you want to run the test, please do the following :
 
`mvn -f testsuite/integration-arquillian/tests/base/pom.xml test -Dtest=org.keycloak.testsuite.webauthn.WebAuthnRegisterAndLoginTest -Dbrowser=chrome  -DchromeArguments="--headless --enable-web-authentication-testing-api"`

I've looked into http://arquillian.org/arquillian-extension-drone/#chrome-options and integration-arquillian-tests-base/src/test/resources/arquillian.xml .

mposolda replied:
Thanks, in my environment it unfortunately didn't help. Still the same behaviour. But thanks for this change anyway.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/pages/webauthn/AppPage.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * @author <a href="mailto:sthorger@redhat.com">Stian Thorgersen</a>
+ */
+public class AppPage extends AbstractPage {
```
This matter is induced by using ChromeDriver not managed by arquillian/drone on RegisterAndLoginTest.java .

As you mentioned testsuite/integration-arquillian/tests/base/pom.xml below, this can be resolved by using ChromeDriver managed by arquillian/drone .

```java
    @FindBy(id = "account")
    private WebElement accountLink;
```
This accountLink is bound to the web element by WebDriver managed by arquillian/drone .
Therefore, it can not be used by RegisterAndLoginTest because it uses ChromeDriver not managed by arquillian/drone.

``` java
    public void openAccount() {
        driver.findElement(By.id("account")).click();
        //accountLink.click();
    }
```
Therefore, I've modified the method above.

mposolda replied:
Yes, +1 for new class and override only openAccount(). Hope we can manage to have ChromeDriver managed by arquillian/drone, but that's not a blocker for this PR and can be a done as a follow-up task IMO.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/webauthn/RegisterAndLoginTest.java:
```diff
@@ -0,0 +1,285 @@
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class RegisterAndLoginTest extends AbstractTestRealmKeycloakTest {
```
Yes, I'll rename RegisterAndLoginTest.java to WebAuthnRegisterAndLoginTest.java .

I'm not sure why this test could not run on your environment. I'm afraid that whether Web Authentication Testing API  works well depends on the platform.

mposolda replied:
Thanks for renaming class.

I will try to do some more investigation about Web Authentication Testing API and why the test doesn't work in my environment. I know that when using Web Authentication Testing API manually with my chrome browser, it works fine.

tnorimat committed:
rename file names

tnorimat commented on services/src/main/java/org/keycloak/models/jpa/converter/AttestationStatementConverter.java:
```diff
@@ -0,0 +1,38 @@
+
+public class AttestationStatementConverter implements AttributeConverter<AttestationStatement, String> {
+
+    private CborConverter converter = new CborConverter(); //TODO: Inject by CDI to make it singleton
```
I'll check whether it can be realized as singleton, and if so, I would like to make it singleton in the way as you said.

mposolda replied:
cool, Thanks.

IMO we can again possibly create follow-up JIRA to have this PR in to be available to wider community for testing earlier?

tnorimat replied:
Yes, I agree with your idea creating follow-up JIRA to be available to wider community.

tnorimat commented on services/src/main/java/org/keycloak/models/jpa/converter/CredentialPublicKeyConverter.java:
```diff
@@ -0,0 +1,38 @@
+
+public class CredentialPublicKeyConverter implements AttributeConverter<CredentialPublicKey, String> {
+
+    private CborConverter converter = new CborConverter(); //TODO: Inject by CDI to make it singleton
```
I'll check whether it can be realized as singleton, and if so, I would like to make it singleton in the way as you said.

tnorimat commented on services/src/main/java/org/keycloak/credential/WebAuthnCredentialProvider.java:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    private void dumpWebAuthnCredentialModel(WebAuthnCredentialModel auth) {
+        logger.debugv("  Context Credential Info::");
```
What do you think about the following?

`WebAuthnCredentialModel`

```java
    public String toString() {
        StringBuffer sb = new StringBuffer();
        if (authenticatorId != null)
            sb.append("Authenticator Id = ")
              .append(authenticatorId)
              .append(",");
        if (attestationStatement != null)
            sb.append("Attestation Statement Format = ")
              .append(attestationStatement.getFormat())
              .append(",");
        if (attestedCredentialData != null) {
            sb.append("AAGUID = ")
              .append(attestedCredentialData.getAaguid().toString())
              .append(",");
            sb.append("CREDENTIAL_ID = ")
              .append(Base64.encodeBytes(attestedCredentialData.getCredentialId()))
              .append(",");
            CredentialPublicKey credPubKey = attestedCredentialData.getCredentialPublicKey();
            byte[] keyId = credPubKey.getKeyId();
            if (keyId != null)
                sb.append("CREDENTIAL_PUBLIC_KEY.key_id = ")
                  .append(Base64.encodeBytes(keyId))
                  .append(",");
            sb.append("CREDENTIAL_PUBLIC_KEY.algorithm = ")
              .append(credPubKey.getAlgorithm().name())
              .append(",");
            sb.append("CREDENTIAL_PUBLIC_KEY.key_type = ")
              .append(credPubKey.getKeyType().name())
              .append(",");
        }
        if (authenticationContext != null) {
            // only set on Authentication
            sb.append("Credential Id = ")
              .append(Base64.encodeBytes(authenticationContext.getCredentialId()))
              .append(",");
        }
        if (sb.length() > 0)
            sb.deleteCharAt(sb.lastIndexOf(","));
        return sb.toString();
    }
```

`WebAuthnCredentialProvider`

```java
        if(logger.isDebugEnabled()) {
            dumpCredentialModel(model);
            dumpWebAuthnCredentialModel(webAuthnModel);
        }

    private void dumpWebAuthnCredentialModel(WebAuthnCredentialModel auth) {
        logger.debug("  Context Credential Info::");
        logger.debug(auth);
    }

```

mposolda replied:
+1 for something like this. Very minor thing: StringBuilder is slightly better option than StringBuffer in this case. See for example https://stackoverflow.com/questions/355089/difference-between-stringbuilder-and-stringbuffer

tnorimat replied:
Thank you for your advice. I'll use StringBuilder instead of StringBuffer.

tnorimat commented on services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnRegister.java:
```diff
@@ -0,0 +1,297 @@
+                .detail("public_key_credential_aaguid", aaguid);
+            context.success();
+        } catch (WebAuthnException wae) {
+            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, wae.getMessage());
```
What do you think about the following?

```java
        } catch (WebAuthnException wae) {
            if (logger.isDebugEnabled()) wae.printStackTrace();
            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, wae.getMessage());
            return;
```

mposolda replied:
I vote for:
```
        } catch (WebAuthnException wae) {
            if (logger.isDebugEnabled()) logger.debug(wae.getMessage(), wae.printStackTrace());
            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, wae.getMessage());
            return;
```

This will mean that exception will be logged to the logger, not to the error output like with "wae.printStackTrace()"

tnorimat replied:
Yes, I agree with you. I'll use logger.debug.

tnorimat commented on services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnRegister.java:
```diff
@@ -0,0 +1,297 @@
+            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, wae.getMessage());
+            return;
+        } catch (Exception e) {
+            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, e.getMessage());
```
what do you think about the follwing?

```java

        } catch (Exception e) {
            if (logger.isDebugEnabled()) e.printStackTrace();
            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, e.getMessage());
            return;
        }

```

mposolda replied:
Same as above

tnorimat replied:
Yes, I agree with you. I'll use logger.debug

tnorimat commented on themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties:
```diff
@@ -1042,6 +1042,31 @@ table-of-password-policies=Table of Password Policies
+webauthn-user-verification-requirement=User Verification Requirement
+webauthn-user-verification-requirement.tooltip=It tells an authenticator confirm actually verifying a user.
+webauthn-create-timeout=Timeout
+webauthn-create-timeout.tooltip=Timeout value for creating user's public key credential
```
Accouding to [WebAuthn API specification](https://www.w3.org/TR/webauthn/#dom-publickeycredentialcreationoptions-timeout), this option is in milliseconds.

However, people are unfamiliar with milliseconds, keycloak and its UI treats it in seconds and convert it when calling navigator.credentials.create().

We can call navigator.credentials.create() without this timeout option because this option is not mandatory according to [WebAuthn API specification](https://www.w3.org/TR/webauthn/#dom-publickeycredentialcreationoptions-timeout). Therefore, I've decided that if this timeout value is 0, keycloak calls navigator.credentials.create() without this timeout option. 

What do you think about the following?

```text
webauthn-create-timeout.tooltip=Timeout value for creating user's public key credential in seconds. If set to 0, this timeout option is not adapted.
```

mposolda replied:
+1 for the tooltip proposal you mentioned.

tnorimat commented on services/src/main/java/org/keycloak/models/jpa/converter/AAGUIDConverter.java:
```diff
@@ -0,0 +1,33 @@
+
+import javax.persistence.AttributeConverter;
+
+public class AAGUIDConverter implements AttributeConverter<AAGUID, byte[]> {
```
I'll try not to use javax.persistence for AAGUIDConverter, AttestationStatementConverter and CredentialPublicKeyConverter and remote JPA dependency in the "services" module.

mposolda replied:
Cool, Thanks! This one will be good to handle in this PR and hopefully won't take much time to fix.

tnorimat replied:
Thank you. I have the idea and try it. However, if it fails or seems to take a lot of time, I'll work in favor of other tasks first.

tnorimat commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnAuthenticator.java:
```diff
@@ -0,0 +1,261 @@
+
+        // read options from authenticator config
+        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
+        String userVerificationRequirement = config.getConfig().get(WebAuthnAuthenticatorFactory.USER_VERIFICATION_REQUIREMENT);
```
Yes, as you pointed out, this NPE happens if WebAuthnAuthenticator's config is not generated.
In such the case, I think the default setting be employed.

What do you think about the following?

```java
        // read options from authenticator config
        AuthenticatorConfigModel config = context.getAuthenticatorConfig();
        String userVerificationRequirement = Constants.DEFAULT_WEBAUTHN_POLICY_NOT_SPECIFIED;
        if (config != null) {
            userVerificationRequirement = config.getConfig().get(WebAuthnAuthenticatorFactory.USER_VERIFICATION_REQUIREMENT);
            if (WebAuthnConstants.OPTION_NOT_SPECIFIED.equals(userVerificationRequirement)) userVerificationRequirement = Constants.DEFAULT_WEBAUTHN_POLICY_NOT_SPECIFIED;
        }
        form.setAttribute(WebAuthnConstants.USER_VERIFICATION, userVerificationRequirement);
```

mposolda replied:
That can work. Or maybe even remove the userVerificationRequirement from authentication config and just use the one from the webauthn policy as I mentioned in the other comment?

tnorimat replied:
OK, I would like to remove the config for WebAuthnAuthenticator and get the userVerificationRequirement from WebAuthnPolicy. What do you think about that?

mposolda replied:
+1, Thanks!

mposolda commented:
> 1. Dependencies (apache-kerby-ans1)

> I hope that apache-kerby-ans1 is inculded at this 1st implementation phase.

> At this 1st implementation phase, apache-kerby-ans1 is not used by the following reasons :

>    apache-kerby-ans1 is used for treating Android Key Attestation Statement TPM Attestation Statement
>    this 1st implementation phase does not verify any type of Attestation Statement

> However, apache-kerby-ans1 is needed in the future when keycloak comes to support verifying > Attestation Statement.
> 2. Dependencies (webauthn4j-utils)

> I hope webauthn4j-utils is inculded at this 1st implementation phase.

> webauthn4j-utils' Base64URL encoder/decoder can be replaced with keycloak's ones. However, this PR uses WebAuthnException defined in webauthn4j-utils. webauthn4j-core and this PR uses WebAuthnException to notify that WebAuthn related error occurs. Also, webauthn4j-core itself uses some classes of webauthn4j-utils.

Thanks for the info. I need to doublecheck dependencies with the team.

For more details, adding every 3rd party dependency is a bit of pain due the non-trivial internal productization process we have. So we're working on doublecheck if everything will be fine in this regard. Every dependency also adds some more potentials for CVE etc. Will get back to you once I have some more info within the team.

> 3. WebAuthn Authenticator Label Edit on User Account Service

> I've fixed to edit the label on the account page of User Account Service.

Thanks

> 4. The way of WebAuthn Authenticator working (similar to OTP Authenticator)

> I see. I've fixed WebAuthn Authenticator to work as you mentioned.

Thanks!

> 5. User Verification Requirement configuration

> I hope UserVerificationRequirement can be set on both WebAuthn authenticator's registration and authentication independently at this time.

> UserVerificationRequirement can be configured independently according to WebAuthn specification.

> If keycloak have one UserVerificationRequirement configuration, the following 3 types of configuration are possible :

>    the same setting of it is applied to both WebAuthn authenticator's registration and authentication (e.g. "required" on registration and "required" on authentication ).

>    the setting of it is applied to only WebAuthn authenticator's registration

>    the setting of it is applied to only WebAuthn authenticator's authentication

> I think that having one UserVerificationRequirement configuration seems to be valid if not any combination of different settings of UserVerificationRequirement (e.g. "required" on registration and "preferred" on authentication) other than the 3 ways of configuration above is meaningful in the real world. I'm not sure that.

I see the point, but my vote is still to rather have single switch at the WebAuthn policy level. I am not sure how common will be the use-case, when there is "required" on registration and "preferred" on authentication or so? Wouldn't it be more common that admin will prefer to have same for both? We can eventually add more switches if we see the demand, but for now, I prefer to remove.

If you strongly feel to keep the switch on the authenticator, I am fine with that. But in that case, it will be good at least to fix the NullPointerException, which just happened to me when I did some manual testing. I've added authenticator to the flow without configuring it and this NPE happened to me due the missing configuration. So will be at least good to fix that one, if you want to keep the "config" on the authenticator:

```
19:43:07,004 WARN  XNIO-1 task-17 [org.keycloak.services] KC-SERVICES0013: Failed authentication
java.lang.NullPointerException
	at org.keycloak.authentication.authenticators.browser.WebAuthnAuthenticator.authenticate(WebAuthnAuthenticator.java:87)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:232)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processAction(DefaultAuthenticationFlow.java:128)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processAction(DefaultAuthenticationFlow.java:99)
```

> 6. Testing

> I've add @ignore onto RegisterAndLoginTest.

Thanks.

In my environment, when I run the test for example with:
```
cd testsuite/integration-arquillian
mvn clean install -Dtest=RegisterAndLoginTest
```
The Chrome doesn't start with headless mode in my case. It is started normally in the GUI. But the Chrome GUI is just stuck after the start and nothing is tested.

I am pretty sure that we can solve it later as in our team we're running chrome in headless mode in some of our tests. We can address when we add the arquillian/drone support to this.

tnorimat commented on services/src/main/java/org/keycloak/authentication/authenticators/browser/WebAuthnAuthenticator.java:
```diff
@@ -0,0 +1,261 @@
+        if (userId == null || userId.isEmpty()) {
+            // Resident Key not supported public key credential was used
+            // so rely on the user that has already been authenticated
+            userId = context.getUser().getId();
```
I think NPE does not happen in the following block :

```java
        // existing User Handle means that the authenticator used Resident Key supported public key credential
        if (userId == null || userId.isEmpty()) {
            // Resident Key not supported public key credential was used
            // so rely on the user that has already been authenticated
            userId = context.getUser().getId();
        } else {
```

The reason is as follows:

When NPE occurs, the follwing two condition need be met :

1. `context.getUser()` is null
   which means that no user is not yet authenticated on keycloak

2. `userId` is null or empty
   which means that `navigator.credentials.get()` returns successful response, but it does not include `User Handle`.

Considering 1, in `WebAutnAuthenticator.authenticate()`, keycloak can not set the `navigator.credentials.get()`'s parameter `PublicKeyCredentialRequestOptions.allowCredentials`.

`Credential ID` has to be set to this, however, keycloak can not determine whose `Credential ID` has to set to it because keycloak does not yet authenticate any user at that time.

Therefore, the browser call `navigator.credentials.get()` without the parameter `PublicKeyCredentialRequestOptions.allowCredentials`.

After calling `navigator.credentials.get()`, in order for keycloak to reach the codes above, the following conditions has to be met :

- `navigator.credentials.get()` returns a successful response
- `userId` is null or empty

To get a successful response by calling `navigator.credentials.get()` without `PublicKeyCredentialRequestOptions.allowCredentials`, the following condition need to be met at least :

- The webauthn authenticator has the capability of `Resident Key`
- The webauthn authenticator has already created the public key credential as `client-side-resident public key credential` on `navigator.credentials.create()`

and, such the webauthn authenticator returns successful response with `User Handle`. (refer to https://www.w3.org/TR/webauthn/#op-make-cred 7-4-4)

Considering all above, if `context.getUser()` is null, `userId` is not null on the codes above.

mposolda replied:
Makes sense, Thanks!

tnorimat commented on services/src/main/java/org/keycloak/credential/WebAuthnCredentialProvider.java:
```diff
@@ -0,0 +1,257 @@
+                    long count = auth.getCount();
+                    auth.setCount(count + 1);
+                    CredentialModel cred = createCredentialModel(auth);
+                    session.userCredentialManager().updateCredential(realm, user, cred);
```
I think the counter need to be updated every time the authentication is successfully carried out. The reason is described in https://www.w3.org/TR/webauthn/#signature-counter and https://www.w3.org/TR/webauthn/#verifying-assertion 17 .

Therefore, I would like not to modify this code, but I agree with your idea creating follow-up JIRA to improve the performance later on.

mposolda replied:
I see, thanks for the explain. This raises few interesting questions:
- Firstly in 6.1.1, it is mentioned that "Authenticators SHOULD implement a signature counter feature." . So implemeting it is not mandatory for authenticators per my understanding. I wonder that when authenticator doesn't support counter and will still return value 0, but credential stored on Keycloak side will have value greater than 0, what will happen during validation? Won't the WebAuthnException be thrown in that case? If yes, I am not sure if it is good behaviour as we should take into account that some authenticators may not support the counters?

- Another point - during validation it is mentioned that:
```
Relying Parties should incorporate this information into their risk scoring. Whether the Relying Party updates the stored signature counter value in this case, or not, or fails the authentication ceremony or not, is Relying Party-specific.
```
I can imagine that some admins may prefer stronger performance and avoid saving info during each credential, so maybe we can add the switch whether to support it or not?

Anyway, +1 to do this later. No need to change anything in this PR. Just wanted to clarify this :)

tnorimat replied:
>Firstly in 6.1.1, it is mentioned that "Authenticators SHOULD implement a signature counter feature." . So implemeting it is not mandatory for authenticators per my understanding. I wonder that when authenticator doesn't support counter and will still return value 0, but credential stored on Keycloak side will have value greater than 0, what will happen during validation? Won't the WebAuthnException be thrown in that case? If yes, I am not sure if it is good behaviour as we should take into account that some authenticators may not support the counters?

It does not happen that keycloak throws WebAuthnException in such the situation. Webauthn4j implemented WebAuthn specification rigorously. 
https://github.com/webauthn4j/webauthn4j/blob/a166a383b038fc2edbb9859f9540cfae1db8ce2a/webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationContextValidator.java#L170

tnorimat replied:
I agree on your idea of adding the switch to control the verification about the counter value in order to avoid frequent write access to DB considering performance. However, I think it can be realized as the other follow-up JIRA. What do you think about it?

mposolda replied:
+1 for adding later. Thanks for pointing the webauthn4j code related to validation.

tnorimat commented on themes/src/main/resources/theme/base/login/webauthn-register.ftl:
```diff
@@ -0,0 +1,136 @@
+
+            });
+
+        function getPubKeyCredParams(signatureAlgorithms) {
```
As for using single bean class, I would like to try it as you said.

As for moving the business logic inside UI template onto java class, I'm afraid it takes a relatively lot of time compared with other tasks.

I've already confirmed it before as follows :

- create the business logic implemented endpoint in the REST manner that receives the user's input about the options for WebAuthn API,  and returns the option itself for WebAuthn API (navigator.credentials.create(), .get())

- create the UI template which receives the user's input and access the above endpoint  in the asynchronous manner, and call WebAuthn API (navigator.credentials.create(), .get()).

However, I'm not sure how to secure this endpoint like other endpoints (e.g. login-actions). Therefore I've decided that this feature has not been included in this PR.

mposolda replied:
Yep, ok. That's fine.

I personally wouldn't try to use REST API for this. At least not in this PR. Binding the parameters to the freemarker is fine IMO.

tnorimat replied:
Yes, at this PR, I'll try to bind parameters to the freemarker as you said.

mposolda replied:
I guess one option could be to bind freemarker objects as strings? For example in java code, you can create the "representation" object and then convert it to JSON String with the KEycloak JsonSerialization.writeValueAsString and bind this string to the freemarker (or addd it as a property to freemarker bean). Then in freemarker template JS code, you can call "JSON.parse" to get back the JSON object.

Just a minor thing, not a blocker for this PR IMO.

tnorimat replied:
I've try it but had some problem. Some special characters in the string processed by JsonSerialization.writeValueAsString has been replaced with HTML character references (e.g. 
 " -> &quot) automatically. Therefore, JSON.parse failed. It seems that keycloak automatically sanitizes any variable binded into Freemarker template.

Maybe the following seems work but it need some conversion work :

[Java]
- The representation of PublicKeyCredentialCreationOptions -> JSON as String
- JSON as String -> Byte Array
- Byte Array -> Base64URL Encoded String

[Javascript]
- Base64URL Encoded Sting -> Byte Array
- Byte Array -> JSON as String
- JSON as String -> Object in Javascript

mposolda replied:
I see. Then maybe we rather keep as it is rather than add more layers of conversion (array of bytes, Base64URL etc), which can also affect performance a bit etc?

tnorimat replied:
Yes, I would like to work on the other task (especially, not using UserModel's attribute for public key credential's metadata) and will back to this issue later on. What do you think about it?

mposolda replied:
+1

tnorimat commented:
@mposolda thank you for your comments. 

> 5. User Verification Requirement configuration

Yes, I agree with you. As mentioned above, I would like to remove the config for WebAuthnAuthenticator and get the userVerificationRequirement from WebAuthnPolicy. We can only use the single switch and avoid NPE of WebAuthnAuthenticator's config. What do you think about that?

tnorimat committed:
improve tooltip message

tnorimat committed:
improve logging

tnorimat committed:
remove javax.persistence dependency from keycloak-services

tnorimat committed:
make converter singleton

tnorimat committed:
avoid NPE of authenticator's config

mposolda commented:
> @mposolda thank you for your comments.
> 
> > 1. User Verification Requirement configuration
> 
> Yes, I agree with you. As mentioned above, I would like to remove the config for WebAuthnAuthenticator and get the userVerificationRequirement from WebAuthnPolicy. We can only use the single switch and avoid NPE of WebAuthnAuthenticator's config. What do you think about that?

+1, Thanks!

tnorimat committed:
remove userVerificationRequirement from WebAuthnAuthenticator's config

tnorimat committed:
use keycloak's Base64URL encoder/decoder

mposolda commented:
@tnorimat So just a summary, that from my PoV. the only thing blocking this PR merge are 3rd party dependencies. We're still evaluating on our side if there won't be any issues with 3rd party dependencies and if it's possible to use them in our internal productization process.

Still, one note regarding apache-kerby-asn1. This one dependency is not needed in the PR as you mentioned. If we want later add the support for verification of Android Key Attestation Statement, is it a chance to use a bouncycastle library for deal with ASN1 format instead of apache-kerby? Bouncycastle has an advantage for us that it is already available in Widlfly and used in some Keycloak functionalities. For example see keycloak class UserIdentityExtractor.

So will be nice to avoid having two libraries for serialize/deserialize ASN1. Depends how tightly coupled is apache-kerby with webauthn4j? Just a note that org.bouncycastle:bcprov-jdk15on:1.60 also doesn't have any 3rd party dependencies similarly like apache-kerby.

tnorimat committed:
remove apache-kerby-asn1 in the first implementation phase

tnorimat commented:
Considering the scope of this 1st implementation phase,  I've removed apache-kerby-asn1. Also, I've removed the option "direct" and "indirect" of AttestationConveyancePreference  in WebAuthn Policy in order to avoid treating Android Key Attestation Statement and TPM Attestation Statement.

Therefore, the scope of this 1st implementation phase should be only internal use (inside the organization, company) because keycloak does not verify Attestation Statement .

tnorimat committed:
use arquillian drone managing chrome driver and its option

mposolda commented:
@tnorimat Thanks for removing apache-kerby-asn1 from the PR. I've asked Yoshikazu Nojima about the possibility of replace apache-kerby with bouncycastle and added you to the CC of the email.

mposolda commented:
@tnorimat Just one more note regarding credential management and credential metadata in attributes (Your Issue 1). I agree that it's not 100% ideal, on the other hand, I probably wouldn't spend too much time on this as there is some refactoring planned in credentials management and it's possible that this part would be refactored. See some details in the other specification: https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md#design-proposal-for-keycloak-9694 .

If you accidentally have some spare time, maybe will be good to have some documentation for the webauthn as well? That seems to me as one of the 2 remaining things blocking this PR.

The second thing are those 3rd party dependencies, but I think you can't do much about it. We're working on it and make sure that productization work on our side. I don't have exact timeline, but will keep you updated. Thanks again for all your work on this PR!

tnorimat commented:
@mposolda I see, till I back to the office, I would like to prepare the documentation. 

As for 3rd party dependencies, I'm afraid I could do little on it as you said.

As for other remaining tasks especially the credential management, I also think that it seems to be heavy to resolve them on this 1st implementation phase PR and appropriate to implement them after realizing other features (Managing multi-factor authentication and Step-up authentication, Application Initiated Actions, Secure Credentials Store, Account Console and Account REST API). 

Therefore, I'll focus on the documentation for the moment.

tnorimat commented:
The documentation PR is https://github.com/keycloak/keycloak-documentation/pull/737 .

tnorimat commented:
@mposolda

Summing up the current situation, it seems the following :

- Issues blocking this PR's merge (2 in total)
  1. Documentation
     I've issued the documentation PR(https://github.com/keycloak/keycloak-documentation/pull/737) and revised the sentences(https://github.com/keycloak/keycloak-documentation/pull/737/commits/423fc4d1342d809c9cf6f5433877e2929c7a18d7) incorporating review comments.
  2. 3rd Party Dependencies
     I can do little on this issue anymore so that I'll wait for the result of your team's internal productization process.
     
Is that correct?

Also, I'll comments on the summed up current situation of "Issues NOT blocking this PR's merge" later on.

mposolda commented:
@tnorimat yes, great summary. Thanks.

Regarding documentation, I've just approved your PR. Andy will do the review too and I don't expect he finds anything major. Hopefully this will be ok soon.

Regarding productization, we're working on it and looks promising. I am not 100% sure, but hope is to have this PR merged next week.

Feel free to work on follow-up tasks if you have time. For example https://issues.jboss.org/browse/KEYCLOAK-11372 is quite safe, as it shouldn't touch the parts, which will be provided by other work related to "Multi-factor and step-up authentication" support. But please don't update this PR, but use different branch for the follow-up work. Will be better to add one thing at a time instead of adding more stuff to this existing PR. I hope it will not cause much headaches... Let me know if you see any issue with it. Thanks

tnorimat commented:
@mposolda 

Thank you for your confirmation.

As you said, I'll look into the follow-up task described in https://issues.jboss.org/browse/KEYCLOAK-11372 . I also think that this task does not depend on any other features mentioned in the related design documents in keycloak-community.

tnorimat committed:
support webauthn4j 0.9.12.RELEASE

tnorimat commented:
The latest webauthn4j (0.9.12.RELEASE) is incorprated.
This implemented the Attestation Statement's serialization/deserialization in the more secure way so that the existing serialized Attestation Statement in the user's webauthn authenticator's credential heretofore can not be deserialized by this commits. Please re-register such the credentials.

mposolda commented:
@tnorimat I am sorry to bother, but is it possible to add the last commit with webauthn4j update to separate branch and remove it from this PR?

We're quite close regarding our productization process to be finished and we're working with webauthn4j 0.9.7. Upgrading to 0.9.12 at this stage would mean that we will need to redo it, which would mean that this PR will be delayed by another few weeks.

I would prefer to have this PR merged ASAP, so that we can work on a follow-up work for multi-factor prototype.

The update to webauthn4j will be good to do later and send separate PR for it. We may need to do our internal productization anyway, but at least the PR for multi-factor prototype won't be blocked. I hope it will be possible to include the webauthn4j update with the PR for KEYCLOAK-11372 and/or KEYCLOAK-11373 as we may need to include apache-kerby-asn1 for that (so the productization work on our side will be needed anyway, and will be good to do it once). Also will be good for the future to have updates to webauthn4j more planned to avoid the productization overhead on our side (EG. not update to every new webauthn4j every week, but rather once in a while when there are more fixes and useful improvements).

tnorimat committed:
Revert "support webauthn4j 0.9.12.RELEASE"

This reverts commit 96da92a9142c9b47be0e35577a59d58c7b1b7f7d.

tnorimat commented:
I see, I've reverted the commit for upgrading webauthn4j to 0.9.12.RELEASE.
I'll try again to plan for this upgrading if this PR is merged.

tnorimat commented:
@mposolda

Summing up the current situation, it seems the following :

- Issues NOT blocking this PR's merge (9 in total)
  1. Arquillian Integration Testing
     Managed the existing ticket [[SPIKE] Investigate how to test Web Authentication](https://issues.jboss.org/browse/KEYCLOAK-9359) .
  2. Support for attestation statement verification
     Managed the existing ticket [Support for attestation statement verification](https://issues.jboss.org/browse/KEYCLOAK-11372) .
  3. Move the logic of constructing WebAuthn API's option to keycloak server side
     I will create the new follow-up ticket for this issue.
  4. Update the version of  WebAuthn4j if including major feature support/bug fix/breaking change
     I will create the new follow-up ticket for this issue.
  5. Manage WebAuthn authenticator's metadata by WebAuthn Credetial Provider
     I will create the new follow-up ticket for this issue.
     This issue can be worked on after Refactoring Credential in [Managing multi-factor authentication and Step-up authentication in Keycloak](https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md#design-proposal-for-keycloak-9694)
  6. Avoid WebAuthn authenticator credential's counter value check and update (performance concern)
     I will create the new follow-up ticket for this issue.
     This issue can be worked on after Refactoring Credential in [Managing multi-factor authentication and Step-up authentication in Keycloak](https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md#design-proposal-for-keycloak-9694)
  7. Cache WebAuthn authenticator's credential (performance concern)
     I will create the new follow-up ticket for this issue.
     This issue can be worked on after Refactoring Credential in [Managing multi-factor authentication and Step-up authentication in Keycloak](https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md#design-proposal-for-keycloak-9694)
  8. Edit WebAuthn authenticator credential's metadata by Admin
     I will create the new ticket for this issue.
     This issue can be worked on after realizing new Account Console and Account REST API.
  9. Delete registered WebAuthn authenticator creadential by User
     I will create the new ticket for this issue.
     This issue can be worked on after realizing new Account Console and Account REST API.

mposolda commented:
@tnorimat Thanks for the summary. Feel free to create JIRAs and please let me know about them. 

I think that for some issues, JIRAs defacto already exists? Like: 
VIII: https://issues.jboss.org/browse/KEYCLOAK-9694
IX: https://issues.jboss.org/browse/KEYCLOAK-9367

Also if you want to work on something, please let me know, so it doesn't clash with the other work already in progress. So far, it looks that issues II, III, IV from your list are fine to work on for you if you have spare time. It will be beneficial as you have great knowledge of webauthn4j.

stianst commented:
I haven't had a chance to review this properly, but one aspect I don't like is the addition of so many config options on the realm. We need something better than that, but I appreciate there is a limitation in Keycloak today.

With the auth flow improvements we need some way of sharing config/policies between actions and authenticators, without them being realm level attributes.

The general idea is that it should be possible to provide an authenticator (with associated actions) as an extension with the same level of integration as if it is added directly to the KC codebase.

There's probably two ways we can achieve the above. Merge this with the intent of refactoring as the auth flow work is ready, or wait with merging this until auth flow is ready. Most important thing though is that it is refactored eventually.

tnorimat commented:
@mposolda 

Thank you for your comments about Issues NOT blocking this PR's merge. I've revised it directly as you suggested.

As you said, II, III, IV of the list seem to be appropriate for me to deal with. At this time, I would like to work on II.

tnorimat commented:
@stianst 

Thank you for your comments. I've confirmed your point mentioning that all webauthn's configuration parameters are set on the realm level.

I think that this issue applies not only webauthn's case but other existing/future configuration parameters set on the realm level.  Therefore, It might be beneficial that the general configuration setting for more fine grain level (e.g. combination of authenticators and actions) are realized. If so, it seems good to manage this issue by the new JIRA ticket separately.

mposolda commented:
@stianst So far, the plan is to merge this PR and then rebase the follow-up prototype by the cloud trust team with the keycloak master (which will already have this PR merged).

I agree that for the future, we can do much better with regards to credential policies etc. Summary of the requirements as I see them:

1) Credential policies shouldn't be defined at realm level

2) In realm, there should be a way to define more credential policies for same type of credential. And admin should have a possibility to tight those credential policies with different authenticators. For example user may be required to authenticate either with one of those two:
- Passwordless with strong WebAuthn authenticator with "userVerification: required"
- 2-factor authentication with password and less strong WebAuthn credential, which doesn't have "userVerification: required"

3) There should be a way to deploy a JAR with new type of credentials, but also new type of credential policies. In other words, we shouldn't have few hardcoded type of credential policies like "password", "OTP", "WebAuthn", but ability to plug an implementation with your own credential and credential policy.

I can imagine the people should be able to deploy single archive with authenticator, required action, credential provider and credential policy types. In other words, it should be possible to completely remove WebAuthn or OTP from Keycloak (including account mgmt and admin console) just by removing few providers.

Also I think that we may need an ability for administrator to create instances of "credential types" with individual credential policies and being able to link them with individual authenticators or actions as described here: https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md#attaching-the-policy-to-the-credential-type . But this part is not yet added in the prototype by Cloud Trust team.

mposolda reviewed (approved):
We made progress on our side with the internal productization process and doublecheck that things are fine on this side. Approving PR

stianst reviewed (approved):
Haven't done a deep review, but LGTM and my concerns has been resolved my Marek's comments on the plans

mposolda commented:
@stianst Thanks!

Cross-dc tests executed locally by @mabartos and passed (they are known to be broken on travis).

mposolda commented:
@tnorimat PR finally merged today. Thanks again for all your hard work on this!

If you want to continue on more follow-up work, please let me know. For now, I assume that you're looking at issues 2, 3, 4 from this comment: https://github.com/keycloak/keycloak/pull/6248 . Is it correct?

tnorimat commented:
@mposolda 

I'm now working on the Issue 2 "Support for attestation statement verification".
And, I'll be able to start Issue 4 "Update the version of WebAuthn4j" with issuing new follow-up JIRA ticket.

Issue 3 "Move the logic of constructing WebAuthn API's option to keycloak server side" is open now.
I would like to treat with the Issue 3 after completion of the Issue 2 and 4. However, if someone can work on the Issue 3, I am very happy.

tnorimat commented:
The referenced pull request just above is not relevant to this PR. Sorry.
